src/otk/annotation.py

```diff
@@ -0,0 +1,253 @@
+        if isinstance(data, AnnotatedDict):
+            ret = {key: cls.deep_dump(value) for key, value in data.items()}
+        elif isinstance(data, AnnotatedList):
+            ret = [cls.deep_dump(item) for item in data]
+        elif isinstance(data, AnnotatedBool):
+            ret = data.value
+        elif isinstance(data, AnnotatedFloat):
+            ret = float(data)
+        elif isinstance(data, AnnotatedInt):
+            ret = int(data)
+        elif isinstance(data, AnnotatedStr):
+            ret = str(data)
+
+        return ret
+
+    def add_source_attributes(self, key_data, prefix=""):
+        line_number = key_data.start_mark.line + 1
+        line_number_end = key_data.end_mark.line + 1
+        column = key_data.start_mark.column + 1
+        column_end = key_data.end_mark.column + 1
+        filename = os.path.relpath(key_data.start_mark.name, self._basedir)
+        self.set_annotation(f"{prefix}src", f"{filename}:{line_number}")
+        self.set_annotation(f"{prefix}filename", filename)
+        self.set_annotation(f"{prefix}line_number", line_number)
+        self.set_annotation(f"{prefix}line_number_end", line_number_end)
+        self.set_annotation(f"{prefix}column", column)
+        self.set_annotation(f"{prefix}column_end", column_end)
+
+    def __add__(self, o):
+        ret = self.deep_convert(super().__add__(o))  # pylint: disable=E1101
+        ret.set_annotations(self.squash_annotations([self, o]))
+        return ret
+
+    def __iadd__(self, o):
+        return self.__add__(o)
+
+    def __radd__(self, o):
+        return self.deep_convert(o).__add__(self)
+
+    def __deepcopy__(self, memo):
+        if isinstance(self, AnnotatedDict):
+            cls = self.__class__
+            result = cls.__new__(cls)
+            memo[id(self)] = result
+
+            for k, v in self.items():
+                result[k] = copy.deepcopy(v, memo)
+
+        elif isinstance(self, AnnotatedList):
+            cls = self.__class__
+            result = cls.__new__(cls)
+            memo[id(self)] = result
+
+            for v in self:
+                result.append(copy.deepcopy(v, memo))
+        elif isinstance(self, (AnnotatedStr, AnnotatedInt, AnnotatedFloat, AnnotatedBool, AnnotatedPath)):
+            result = copy.copy(self)
+        else:
+            result = None  # just for the linter - that should never happen
+
+        result.set_annotations(self.get_annotations().copy())
+
+        return result
+
+    @classmethod
+    def squash_annotations(cls, annotation_list: list) -> dict:
+        """ just aggregates all annotations in the list.
+
+        new annotation keys get appended to the resulting dict
+
+        duplicate annotations will get converted to string
+        (if you need an exception for a key please implement here)
+
+        for "src" there is an exception
+        this is expected to be in the form "filename:line_number"
+        and will be sanely unified
+        """
```

Comment:
This is a rather poor docstring for a method that isn't very obvious or easy to read.

The first like starting with "just" doesn't help.  "Aggregates all annotations in the list" would be a good first line for it, the "just" makes it sound like things will be simple, but the function body suggests otherwise.  But I would also add "into a dictionary keyed by <something>".

"New annotation keys get appended to the dict".  Isn't every key here new?  How can there be old annotation keys?  This looks like a pure function (it's annotated as a classmethod but it's actually static right?  `cls` isn't used).

"Duplicate annotations will get converted to a string".  Does this mean unique annotations are different types?  What does that mean for the result?

"(if you need an exception for a key please implement here)".  I would change this to:
"Exceptions for specific keys are implemented here."

and then follow it with:
```
Key exceptions:
- src: ...
```

But also, I'm not entirely clear what the standard behaviour is like so I don't really understand what the exception is excepting.